sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/MetricStorageRegistry.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.sdk.metrics.internal.state;
+
+import io.opentelemetry.sdk.metrics.internal.descriptor.MetricDescriptor;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * Responsible for storing metrics (by name) and returning access to input pipeline for instrument
+ * wiring.
+ *
+ * <p>The rules of the registry:
+ *
+ * <ul>
+ *   <li>Only one storage type may be registered per-name. Repeated look-ups per-name will return
+ *       the same storage.
+ *   <li>The metric descriptor should be "compatible", when returning an existing metric storage,
+ *       i.e. same type of metric, same name, description etc.
+ *   <li>The registered storage type MUST be either always Asynchronous or always Synchronous. No
+ *       mixing and matching.
+ * </ul>
+ */
+public class MetricStorageRegistry {
+  private final ConcurrentMap<String, MetricStorage> registry = new ConcurrentHashMap<>();
+
+  /**
+   * Returns a {@code Collection} view of the registered {@link MetricStorage}.
+   *
+   * @return a {@code Collection} view of the registered {@link MetricStorage}.
+   */
+  public Collection<MetricStorage> getMetrics() {
+    return Collections.unmodifiableCollection(new ArrayList<>(registry.values()));
+  }
+
+  /**
+   * Registers the given {@code Metric} to this registry. Returns the registered storage if no other
+   * metric with the same name is registered or a previously registered metric with same name and
+   * equal with the current metric, otherwise throws an exception.
+   *
+   * @param storage the metric storage to use or discard.
+   * @return the given metric storage if no metric with same name already registered, otherwise the
+   *     previous registered instrument.
+   * @throws IllegalArgumentException if instrument cannot be registered.
+   */
+  @SuppressWarnings("unchecked")
+  public <I extends MetricStorage> I register(I storage) {
+    MetricDescriptor descriptor = storage.getMetricDescriptor();
+    MetricStorage oldOrNewStorage =
+        registry.computeIfAbsent(descriptor.getName().toLowerCase(), key -> storage);
+    // Make sure the storage is compatible.
+    if (!oldOrNewStorage.getMetricDescriptor().isCompatibleWith(descriptor)) {
+      throw new IllegalArgumentException(
+          "Metric with same name and different descriptor already created.   Found: "
```

Comment:
(Putting the comment in the source where it belongs). This message shouldn't say "Metric with same...", because there isn't a thing called a "Metric" here. It's the Instrument that is trying to be created with the same name. This message will be confusing to users to see it in their code, IMO.

Reply:
Well, except Views make this hella confusing.  Specifically this *COULD* be a name conflict from an instrument to a view that-has-a-rename.

I agree it's confusing to users.   I decided to consolidate on this notion of Metric.   When updating the View implementation I was going to add some provenance here so we can give good error messages.

Specifically, right now you see it lists the metric descriptors.   Post-View-Rework I expect this to have logic somewhat as follows:


1. If conflict involves two instruments with default aggregations, error as instrument conflict.
2. If conflict involves one instrument + one view re-name, error as instrument conflicting with "view" metric output.  Output the name of the view + its selector.
3. Conflicts involving multiple views *should* be caught when building the SDK and shouldn't make it to this level.


I can take a crack at showing that logic here or keeping it instrument based, since you can't have view renames just yet.

Reply:
yeah...that will get even more confusing. I won't hold up this PR for it... and getting some actual user feedback will be valuable.

Reply:
+100